content/2017-11-05_code-review.md

```diff
@@ -0,0 +1,78 @@
+
+Jedna z věcí, které jsou od počátku zcela zakořeněny v softwarové kultuře [Apiary](https://apiary.io/), jsou právě code reviews. Natolik, že pokud se vyjímečně stane, že začleňuji kód, který review nedostal, cítím se stejně provinile, jako bych použil `goto`, `eval`, nebo jinou věc, která programátory od mala budí ze spaní. Jak to tedy funguje u nás?
+
+## Kdo kontroluje
+
+Když nás bylo málo a neměli jsme firmu rozdělenou na týmy, dělal každý zrovna to, co bylo potřeba. Každý rozuměl každé části produktu, a uměl komukoliv udělat code review. Velmi brzy po tom, co jsem nastoupil, jsem dělal review ostříleným mazákům a oni mě. V code review neexistovala žádná hierarchie, která by říkala, že zkušenější dělají review méně zkušeným. Díky tomu se znalost codebase a schopnost code review hezky rozprostřela. Také nikdo neměl pocit, že by byl chytřejší než ostatní, nebo neomylný. Rovní s rovnými. Ono totiž pokud vám dělá review začátečník a nepochopí váš kód, není to jeho chyba. Napsali jste příliš komplexní a nepochopitelný kód. Až tam bude chvíli ležet a vy se k němu po čase vrátíte, budete ho také číst jako začátečník.
+
+Když jsme vytvořili týmy kolem produktových oblastí ([_bounded contexts_](https://en.wikipedia.org/wiki/Domain-driven_design#Bounded_context)), změnilo se akorát to, že reviews si dělají lidé převážně v rámci týmu.
+
+## Co se kontroluje
+
+Code review se dělá zcela na všem, co má být začleněno do hlavní větve. I ve chvíli, kdy jsem ve tři ráno musel v rámci služby vstát a opravit produkci kvůli nějakému průšvihu, nechal jsem si udělat code review. Samozřejmě ne zcela vždy (někdy na to prostě nebyl čas) a také pomůže, když máte i ve tři polovinu firmy vzhůru, protože jsou v San Franciscu :-D
+
+Neexistuje, že bych začal novou knihovnu na zelené louce a tvořil bych to tam sám, bez code review. Kdo pohlídá, že ji navrhuji správně? Kdo bude schopen ji udržovat, když pojedu na dovolenou? Díky code review jsou zaručeny alespoň základní kvality kódu. A také existují minimálně dva lidé ve firmě, kteří danému kódu rozumí.
+
+## Kdy se kontroluje
+
+Toto nemáme úplně pevně dáno, ale slušností je udělat code review co nejdřív, kdy má dotyčný Věnceslav kousek volného času. Nevytrháváme se tedy z programátorského _flow_, ale v zásadě se očekává, že review proběhne aspoň do 24 hodin. Jestliže se totiž na review čeká dlouho, začně to příliš blokovat Uršulu. Někde (zdroj jsem nedohledal) jsem četl, že reviews by se měly dělat ihned, protože neblokovat někoho, kdo má práci hotovou, má vyšší hodnotu než chodit po špičkách kolem _flow_ někoho, kdo programuje. Svým způsobem mi to smysl dává.
+
+## O čem se (ne)diskutuje
+
+Zde záměrně smíchám to, co ve firmě děláme všichni, a to, co dělám já. Nebo co bych si přál, aby bylo běžné. Když přijdu k PR jako někdo, kdo má dělat review, postupuji následovně:
+
+- Přečtu si popis PR. Pokud jej nemá nebo nevysvětluje _proč_ jsou změny prováděny, PR zavřu, nezajímá mě.
```

Comment:
Please explain what "PR zavřu" means - it is not clear to me.
Different tools use different terminologies, but "approve" (schválit) and "decline" (zamítnout) is some sort of industry standard and clear to anybody. But the term "Close" can mean either letting the code to go to the master branch or the opposite.

Reply:
Asi můžeme klidně česky. Zavřením jsem myslel tlačítko <kbd>Close</kbd>. Dělali jsme to tak, než se GitHub naučil ty dnešní reviews. Dnes tedy už PR nezavíráme, asi jedině pokud jsou již překonané nebo zcela chybné. Stará terminologie mi ale očividně vlezla pod kůži. To, co popisuji, je spíše zamítnutí (s komentářem proč a s možností nápravy).

Reply:
Díky za připomínku, opravil jsem to - https://github.com/honzajavorek/honzajavorek.cz/pull/34 Jen pro upřesnění, co jsem přesně myslel tím tlačítkem <kbd>Close</kbd>:

<img width="474" alt="screen shot 2017-11-08 at 20 34 53" src="https://user-images.githubusercontent.com/283441/32570496-8e6a8df2-c4c4-11e7-92fc-600582255bfc.png">

Reply:
Už rozumím, díky.